src/Control/Controller.php

```diff
@@ -690,14 +697,31 @@ public static function join_links($arg = null)
             }
             if ((is_string($arg) && $arg) || is_numeric($arg)) {
                 $arg = (string) $arg;
-                if ($result && substr($result ?? '', -1) != '/' && $arg[0] != '/') {
-                    $result .= "/$arg";
+                if (strlen($result) == 0) {
+                    $result .= $arg;
+                } elseif (substr($result, -1) == '/') {
+                    $result .= trim($arg, '/');
                 } else {
-                    $result .= (substr($result ?? '', -1) == '/' && $arg[0] == '/') ? ltrim($arg, '/') : $arg;
+                    $result .= '/' . trim($arg, '/');
                 }
             }
         }
 
+        // add trailing slash if enabled and url does not contain a file extension
+        if (static::config()->get('add_trailing_slash') === true
+            && substr($result, -1) !== '/'
+            && !preg_match('/^(.*)\.([A-Za-z][A-Za-z0-9]*)$/', $result)
+        ) {
+            $result = rtrim($result, '/') . '/';
```

Comment:
We don't want a trailing slash after an anchor either - if we have something like `/my-page#my-anchor` it should be rewritten as `/my-page/#my-anchor` but I think this code would make it `/my-page/#my-anchor/`
Same with URLs that have parameters on the end e.g. `/my-page?my-var=my-value`

Reply:
This section here only handles the path. Fragment identifiers and query args are added back later.